ui/src/app/utils/subscribable.spec.tsx

```diff
@@ -0,0 +1,126 @@
+import { atom } from './subscribable';
+
+export const delay = ms => {
+  return new Promise(resolve => setTimeout(resolve, ms));
+};
+
+const promiseComplete = () => {
+  let resolveHandler;
+  const promise = new Promise<void>((resolve) => {
+    resolveHandler = value => value && resolve();
+  });
+  return {promise, resolveHandler};
+};
+
+const makeSubscribeFn = (assertions): [() => Promise<void>, (newValue: any, oldValue: any) => Promise<{}>] => {
+  const {promise, resolveHandler} = promiseComplete();
+  const promiseHandler = (): Promise<void> => promise;
+
+  const subscribeFn = (newValue, oldValue) => new Promise((resolve) => {
+    assertions.forEach(assertion => assertion(newValue, oldValue));
+    resolveHandler(true);
+    resolve();
+  });
+
+  return [promiseHandler, subscribeFn];
+};
+
+describe('atom', () => {
+
+  it('should initialize with a value', () => {
+    const testAtom = atom({value: 1});
+    expect(testAtom.get()).toEqual({value: 1});
+  });
+
+  it('should set the value', () => {
+    const testAtom = atom({value: 1});
+    testAtom.set({value: 5});
+    expect(testAtom.get()).toEqual({value: 5});
+  });
+
+  it('should call a single subscribed function', async() => {
+    const testAtom = atom({value: 1});
+    const [complete, subscribeFn] = makeSubscribeFn([
+      newValue => expect(newValue).toEqual({value: 10}),
+      ({}, oldValue) => expect(oldValue).toEqual({value: 20})
+    ]);
+
+    testAtom.set({value: 20});
+    testAtom.subscribe(subscribeFn);
+    testAtom.set({value: 10});
+
+    await complete();
+  });
+
+  it('should call a multiple subscribed functions with new and old values', async() => {
+    const testAtom = atom({value: 1});
+    const assertions = [
+      newValue => expect(newValue).toEqual({value: 10}),
+      ({}, oldValue) => expect(oldValue).toEqual({value: 20})
+    ];
+
+    const [complete1, subscribeFn1] = makeSubscribeFn(assertions);
+    const [complete2, subscribeFn2] = makeSubscribeFn(assertions);
+    const [complete3, subscribeFn3] = makeSubscribeFn(assertions);
+
+    testAtom.set({value: 20});
+    testAtom.subscribe(subscribeFn1);
+    testAtom.subscribe(subscribeFn2);
+    testAtom.subscribe(subscribeFn3);
+    testAtom.set({value: 10});
+
+    await complete1();
+    await complete2();
+    await complete3();
+  });
+
+  it('should not call an unsubscribed function', async() => {
+    const testAtom = atom({value: 1});
+    let value = 1;
```

Comment:
nit: for the purposes of counting the invocations, I think starting with 0 would be more readable

Reply:
done